pkg/util/kubernetes/apiserver/store.go

```diff
@@ -0,0 +1,89 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build kubeapiserver
+
+package apiserver
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/patrickmn/go-cache"
+)
+
+// metaBundleStore is a cache for MetadataMapperBundles for each node in the cluster
+// and allows multiple goroutines to safely get or create meta bundles for the same nodes
+// without overwriting each other.
+type metaBundleStore struct {
+	mu      sync.RWMutex
+	cache   *cache.Cache
+	keyFunc func(...string) string
+}
+
+func (m *metaBundleStore) get(nodeName string) (*MetadataMapperBundle, bool) {
+	cacheKey := m.keyFunc(metadataMapperCachePrefix, nodeName)
+
+	var metaBundle *MetadataMapperBundle
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	v, ok := m.cache.Get(cacheKey)
+	if !ok {
+		return nil, false
+	}
+
+	metaBundle, ok = v.(*MetadataMapperBundle)
+	if !ok {
+		log.Errorf("invalid cache format for the cacheKey: %s", cacheKey)
+		return nil, false
+	}
+
+	return metaBundle, true
+}
+
+func (m *metaBundleStore) getOrCreate(nodeName string) *MetadataMapperBundle {
+	cacheKey := m.keyFunc(metadataMapperCachePrefix, nodeName)
+
+	var metaBundle *MetadataMapperBundle
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	v, ok := m.cache.Get(cacheKey)
+	if ok {
+		metaBundle, ok := v.(*MetadataMapperBundle)
+		if ok {
+			return metaBundle
+		}
+		log.Errorf("invalid cache format for the cacheKey: %s", cacheKey)
+	}
+
+	metaBundle = newMetadataMapperBundle()
+
+	m.cache.Set(cacheKey, metaBundle, cache.NoExpiration)
+
+	return metaBundle
+}
+
+func (m *metaBundleStore) set(nodeName string, metaBundle *MetadataMapperBundle) {
+	cacheKey := m.keyFunc(metadataMapperCachePrefix, nodeName)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.cache.Set(cacheKey, metaBundle, cache.NoExpiration)
```

Comment:
why wouldn't you expire the bundle ?

Reply:
Discussed IRL - There is a logic to remove the bundle if a node is deleted.
We don't want to expire because we want to avoid a case where a node is detected with an empty map and one where it has not been computed yet (difference from the nodelisterner and the endpointlistener).